File: src/common/py/api/ping_command.py
from ..core.messaging import (
    Command,
    CommandReply,
    Message,
    Channel,
    CommandDoneCallback,
    CommandFailCallback,
)
from ..core.messaging.handlers import MessageEmitter


@Message.define("command/general/ping")
class PingCommand(Command):
    """
    A generic PING command.

    Notes:
        Requires a ``PingReply`` reply.
    """

    payload: str = "PING"

    @staticmethod
    def emit(
        message_emitter: MessageEmitter,
        target: Channel,
        *,
        done_callback: CommandDoneCallback | None = None,
        fail_callback: CommandFailCallback | None = None,
        async_callbacks: bool = False,
        timeout: float = 0.0,
        chain: Message | None = None
    ) -> None:
        """
        Helper function to easily emit this message.
        """
        message_emitter.emit_command(
            PingCommand,
            target,
            done_callback=done_callback,
            fail_callback=fail_callback,
            async_callbacks=async_callbacks,
            timeout=timeout,
            chain=chain,
        )


@Message.define("command/general/ping/reply")
class PingReply(CommandReply):
    """
    Reply to the ``PingCommand``.
    """

    payload: str = "PONG"

    @staticmethod
    def emit(
        message_emitter: MessageEmitter,
        cmd: PingCommand,
        success: bool = True,
        message: str = "",
    ):
        """
        Helper function to easily emit this message.
        """
        message_emitter.emit_reply(PingReply, cmd, success=success, message=message)

File: src/common/web/api/PingCommand.ts
import { Command } from "../core/messaging/Command";
import { CommandReply } from "../core/messaging/CommandReply";
import { Message } from "../core/messaging/Message";

/**
 * A generic PING command. Requires a ``PingReply`` reply.
 */
@Message.define("command/general/ping")
export class PingCommand extends Command {
    public readonly payload: string = "PING";
}

/**
 * Reply to the ``PingCommand``.
 */
@Message.define("command/general/ping/reply")
export class PingReply extends CommandReply {
    public readonly payload: string = "PONG";
}
